File: apps/api/app/websockets/events.py
from typing import Any, Literal, Union
from pydantic import BaseModel


class GameStartedEvent(BaseModel):
    type: Literal["game:started"] = "game:started"
    game_id: str
    config: dict[str, Any]


class GameEndedEvent(BaseModel):
    type: Literal["game:ended"] = "game:ended"
    game_id: str
    scoreboard: list[dict[str, Any]]


class PlayerJoinedEvent(BaseModel):
    type: Literal["player:joined"] = "player:joined"
    player_id: str
    name: str
    emoji: str
    game_id: str


class PlayerLeftEvent(BaseModel):
    type: Literal["player:left"] = "player:left"
    player_id: str
    game_id: str


class LocationUpdateEvent(BaseModel):
    type: Literal["location:update"] = "location:update"
    player_id: str
    lat: float
    lng: float
    ts: int


class CircleShrinkEvent(BaseModel):
    type: Literal["circle:shrink"] = "circle:shrink"
    lat: float
    lng: float
    radius_m: float
    next_shrink_at: int


class ChickenAlertEvent(BaseModel):
    type: Literal["chicken:alert"] = "chicken:alert"
    distance_m: float
    team_id: str


class ChickenFoundEvent(BaseModel):
    type: Literal["chicken:found"] = "chicken:found"
    team_id: str
    team_name: str
    order: int


class ChallengeNewEvent(BaseModel):
    type: Literal["challenge:new"] = "challenge:new"
    challenge_id: str
    team_id: str
    title_en: str
    title_fr: str
    points: int
    time_limit_sec: int


class ChallengeSubmittedEvent(BaseModel):
    type: Literal["challenge:submitted"] = "challenge:submitted"
    submission_id: str
    team_id: str
    challenge_id: str
    media_url: str | None


class ChallengeScoredEvent(BaseModel):
    type: Literal["challenge:scored"] = "challenge:scored"
    submission_id: str
    team_id: str
    points: int
    approved: bool


class WeaponFiredEvent(BaseModel):
    type: Literal["weapon:fired"] = "weapon:fired"
    weapon: str
    by_team_id: str
    target_team_id: str | None


class WeaponHitEvent(BaseModel):
    type: Literal["weapon:hit"] = "weapon:hit"
    weapon: str
    effect: str
    target_team_id: str | None


class BarMarkedEvent(BaseModel):
    type: Literal["bar:marked"] = "bar:marked"
    bar_id: str
    team_id: str
    found: bool


GameEvent = Union[
    GameStartedEvent, GameEndedEvent, PlayerJoinedEvent, PlayerLeftEvent,
    LocationUpdateEvent, CircleShrinkEvent, ChickenAlertEvent, ChickenFoundEvent,
    ChallengeNewEvent, ChallengeSubmittedEvent, ChallengeScoredEvent,
    WeaponFiredEvent, WeaponHitEvent, BarMarkedEvent,
]

File: packages/shared/src/types/events.ts
import type { GameConfig } from "./game";

export interface TeamScore {
  teamId: string;
  teamName: string;
  score: number;
  foundOrder: number | null;
}

export type WeaponType = "air_strike" | "spy" | "booby_trap" | "steal" | "decoy" | "silence";

export type GameEvent =
  | { type: "game:started"; config: GameConfig }
  | { type: "game:ended"; scoreboard: TeamScore[] }
  | { type: "player:joined"; playerId: string; name: string; emoji: string; gameId: string }
  | { type: "player:left"; playerId: string; gameId: string }
  | { type: "location:update"; playerId: string; lat: number; lng: number; ts: number }
  | { type: "circle:shrink"; lat: number; lng: number; radiusM: number; nextShrinkAt: number }
  | { type: "chicken:alert"; distanceM: number; teamId: string }
  | { type: "chicken:found"; teamId: string; teamName: string; order: number }
  | { type: "challenge:new"; challengeId: string; teamId: string; titleEn: string; titleFr: string; points: number; timeLimitSec: number }
  | { type: "challenge:submitted"; submissionId: string; teamId: string; challengeId: string; mediaUrl: string | null }
  | { type: "challenge:scored"; submissionId: string; teamId: string; points: number; approved: boolean }
  | { type: "weapon:fired"; weapon: WeaponType; byTeamId: string; targetTeamId: string | null }
  | { type: "weapon:hit"; weapon: WeaponType; effect: string; targetTeamId: string | null }
  | { type: "bar:marked"; barId: string; teamId: string; found: boolean };
